File: kotlin/tutorials/tour-of-orchestration-kotlin/src/main/kotlin/my/example/objects/service.kt
import dev.restate.sdk.ObjectContext
import dev.restate.sdk.SharedObjectContext
import dev.restate.sdk.annotation.Handler
import dev.restate.sdk.annotation.Shared
import dev.restate.sdk.annotation.VirtualObject
import dev.restate.sdk.common.StateKey
import dev.restate.sdk.http.vertx.RestateHttpServer
import dev.restate.sdk.kotlin.endpoint
import dev.restate.serde.TypeRef
import java.time.Instant

@VirtualObject
class UserSubscriptions {
    
    companion object {
        private val SUBSCRIPTIONS = StateKey.of("subscriptions", object : TypeRef<List<String>>() {})
        private val LAST_UPDATED = StateKey.of("lastUpdated", String::class.java)
    }

    @Handler
    suspend fun add(ctx: ObjectContext, subscription: String) {
        // Get current subscriptions
        val subscriptions = ctx.get(SUBSCRIPTIONS).orElse(emptyList()).toMutableList()

        // Add new subscription
        if (!subscriptions.contains(subscription)) {
            subscriptions.add(subscription)
        }
        ctx.set(SUBSCRIPTIONS, subscriptions)

        // Update metrics
        ctx.set(LAST_UPDATED, Instant.now().toString())
    }

    @Shared
    suspend fun getSubscriptions(ctx: SharedObjectContext): List<String> {
        return ctx.get(SUBSCRIPTIONS).orElse(emptyList())
    }
}

fun main() {
    RestateHttpServer.listen(endpoint {
        bind(UserSubscriptions())
    })
}

File: python/tutorials/tour-of-orchestration-python/app/objects/service.py
import restate
from datetime import datetime
from typing import List


user_subscriptions = restate.VirtualObject("UserSubscriptions")


@user_subscriptions.handler()
async def add(ctx: restate.ObjectContext, subscription: str) -> None:
    # Get current subscriptions
    subscriptions = await ctx.get("subscriptions", type_hint=List[str]) or []

    # Add new subscription
    if subscription not in subscriptions:
        subscriptions.append(subscription)

    ctx.set("subscriptions", subscriptions)

    # Update metrics
    ctx.set("lastUpdated", datetime.now().isoformat())


@user_subscriptions.handler("getSubscriptions")
async def get_subscriptions(ctx: restate.ObjectSharedContext) -> List[str]:
    return await ctx.get("subscriptions", type_hint=List[str]) or []
